Blog Article - Change Data Capture with Apache Airflow Part 1 (#22)
astronomer/astro-blog: posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

## Overview

The scheduled approach outlined in this post - enabling change data to be updated in near-real time - has the added benefit of built-in data partitioning formats directly from the process. In a distributed data network, using Airflow as the scheduled approach pushes it into the realm of Consistency and High Availability (CA) in the CAP theorem. That is, you can base SLAs on this approach.

Those capturing data in the format of time-series or events will benefit the most with this approach, but it can be used for any data by adding soft deletes to your transactions and defining the time scope to which changes must be recorded. Eventual consistency is achieved in a robust manner by creating an idempotent approach to how the data is extracted and moved. 

> We'll outline all steps in the process from end to end, from setup to extract. In totality, this includes topics of:

Reviewer: In this series, we'll outline...

Reviewer: In totality, this will include:

- Setting up security and IAM in GCP
- Calling the CloudSQL Export API
- Setting up Airflow and connecting it to GCP
- Building a DAG in Airflow
- Configuring Watermarking for data extract intervals
- Writing an operator in Airflow to extract the data
- Integrating the operator into the DAG
- Scheduling and coordinating the extract
- Deployment to Production on Astronomer Cloud

In a later post, I will show you how to transform or clean the data then add it to a warehouse as part of the same workflow. We will be using the [Astronomer Platform](https://www.astronomer.io/), as it is the easiest way to get your Apache Airflow instance up and running.

## How it Works

The base method for extract uses the [CloudSQL Export API](https://cloud.google.com/sql/docs/postgres/import-export/exporting#rest-v1beta4). This runs an operation with a SQL query, sending the results to a Google Cloud Storage (GCS) bucket as declared by you. You could just as easily extract directly from your database if your system and workload allows for such. 

> Using the extract API provides a robust and easily configurable security layer controlled by a service account, such that no system outside of CloudSQL is touching your production database. It also provides an option for a "serverless offload" option in the case that you need to guarantee zero load on your database from the extract. 

Reviewer: Remove the first "option" here, sounds redundant.

"It also providers a "serverless offload" option".

Author: good catch

> The caveat for the export API in general is you [cannot run concurrent CloudSQL Operations](https://cloud.google.com/sql/docs/postgres/import-export/exporting#too-long-rca), so you want to run each query in serial, a scenario perfectly handled by using pools in Airflow.

Reviewer: Add "that" after is- "is that you cannot..."

Reviewer: "so you will need to run each query in serial.."

In general, the API call looks like this:

```bash
POST https://www.googleapis.com/sql/v1beta4/projects/<project-id>/instances/<instance-id>/export
{
 "exportContext":
   {
      "fileType": "CSV",
      "uri": "gs://bucket_name/path_to_csv_file",
      "databases": ["database_name"],
      "offload": false
      "csvExportOptions":
       {
           "selectQuery":"SELECT Query - no CTE allowed - subquery okay"
       }
   }
}
```

This will query `database_name` (only one allowed at a time) using the enclosed `selectQuery` and format the results to a CSV at the GCS bucket and keypath set in `uri`. 

>   > *Note: This CSV will not have headers, and in turn, you will also run a second operation to obtain the table schema, which will be described below*

Reviewer: "This CSV will not have headers, so you will need to run a second operation..."

## Step 1: Set Up Airflow

The [Astronomer CLI](https://github.com/astronomer/astro-cli) is the easiest way to run Apache Airflow on your machine. From the CLI, you can establish a local testing environment regardless of where you'll be deploying to from there.

> There are two ways to install any version of the Astronomer CLI:

Reviewer: Think we can cut out a lot of this and just link out to the CLI install doc- feels like all of this content distracts from the core messaging of the post and can be offloaded. I'd remove this line down and just add:

To install the CLI, follow our [install guides](link).

Author: Delete the whole section? I added it in the teaching that people clicking away from the page can be bad. What about, "If you already have the Astronomer CLI installed, you can skip to Step ___"

Reviewer 2: I think we can do something in between. For instance, we could just include the Homebrew command for the latest version and say "For alternative installation methods, read our install guide." We then jump straight to the command that confirms the install was successful.

Reviewer: Yeah, this ^ works. Just want to get to the good stuff as quickly as possible.

Author: Went for a balance between @jwitz and @petedejoy here

![Screen Shot 2021-01-21 at 1 13 39 PM](https://user-images.githubusercontent.com/307956/105413253-b27ddd00-5bea-11eb-9b30-81a1c0c8565e.png)

- cURL
- [Homebrew](https://brew.sh/)

> ### **Prerequisites**

Author: Oh, haha. I copy-pasted this from the website into notion. Got it. We should take note of that when repro-ing content.

```python
brew install astronomer/tap/astro@0.16.1
```

### **Install with cURL**

To install the latest version of the Astronomer CLI, run:

```bash
curl -sSL https://install.astronomer.io | sudo bash
```

To install a specific version of the Astronomer CLI, specify `-s -- major.minor.patch` as a flag at the end of the cURL command. To install v0.16.1, for example, run:

```bash
curl -sSL https://install.astronomer.io | sudo bash -s -- v0.16.1
```

### **Confirm the Install**

To make sure that you have the Astronomer CLI installed on your machine, run:

```bash
astro version
```

If the installation was successful, you should see the version of the CLI that you installed in the output:

```bash
Astro CLI Version: 0.15.0
Git Commit: c4fdeda96501ac9b1f3526c97a1c5c9b3f890d71
```

For a breakdown of subcommands and corresponding descriptions, you can always run `$ astro` or `$ astro --help`.

  >*For more information on installing the Astronomer CLI, please check out the page the [Astronomer CLI Quickstart](https://www.astronomer.io/docs/cloud/stable/develop/cli-quickstart).*

## Step 2: GCP Access - CloudSQL to GCS

> To provide access between Airflow, CloudSQL, and a GCS bucket, you first need to give your CloudSQL instance the necessary permissions to make the export to a GCS Bucket, followed by providing Airflow the access to make API calls. 

Reviewer: "To allow for access between Airflow, CloudSQL, and your GCS bucket, you'll need to both give your CloudSQL instance the necessary permissions to export to the GCS bucket and give Airflow access to make API calls to both services."

    ```bash
    $ gcloud config set project your-prod-project-id
    Updated property [core/project]

    $ gcloud sql instances describe your-database-instance-id
    ...stuff
    serviceAccountEmailAddress: <stuff>@gcp-sa-cloud-sql.iam.gserviceaccount.com
    ```

3. Open your GCP Storage Browser and navigate to your destination GCS bucket, or create one, and `Edit Bucket Permissions`. 
<span style="display:block;text-align:center;">
  <img src="../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-19_at_1.38.19_PM.png" style="max-width:50%;">
</span>

4. Take the value for `serviceAccountEmailAddress` and add it as `Storage Object Admin` to your  GCS bucket. 
<span style="display:block;text-align:center;">
  <img src="../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-13_at_3.57.37_PM.png" style="float:left;max-width:50%;">
  <img src="../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-13_at_3.56.24_PM.png" style="float:left;max-width:50%;">
</span><br clear="all">

## Step 3: GCP Access - CloudSQL Export API Service Account

> Next, you need to facilitate CloudSQL API access for your Airflow Instance by creating a service account to be used in Airflow. To do so: 

Reviewer: You'll* just to standardize conjugation across all steps

    ```

3. Next, In the `dags` directory create a new python file with appropriate naming of your choice, such as `change_data_capture.py`. Now (with docker running) start up airflow from the CLI by running:

    ```bash
    astro dev start
    ```

4. Once the containers are built and the webserver starts, you can access the UI at  `[localhost:8080](http://localhost:8080)` and login with `admin:admin`.

## Step 5: Create a Connection to GCP

Now we need to add the service account JSON from GCP to Airflow to create a connection. To do so: 

1. In the Airflow UI, open the `Admin -> Connections` tab and create a new connection with the `Connection Type` of `Google Cloud Platform`. 
2. In the `Scopes` field, add the the following scope:

    ```bash
    https://www.googleapis.com/auth/cloud-platform
    ```

3. In the `Keyfile JSON` field, paste the Keyfile JSON you copied from Step 2. 

    ![](../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-13_at_4.25.24_PM.png)

Keep the `Conn Id` handy. You will use that in your Airflow DAG to reference this connection. When you're ready, save this connection to finalize your access management. 

## Step 6: Configure Pooling

> The next configuration you'll set is a pool. From the [Airflow docs](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=pool#pools), `Airflow pools can be used to limit the execution parallelism on arbitrary sets of tasks`, and therefore, using them will enable you to create other parallel jobs while still requiring the export API tasks to be guaranteed to be run in serial. 

Reviewer: Would change the end of this up a bit to be less verbose:

"...will enable you to create other parallel jobs while still guaranteeing that your export API tasks will run in serial."

> To create this is simple. Just navigate in the top bar to `Admin -> Pools`, then hit the `+` sign to add a new record. Choose a pool name and description to your liking, but choose the number of slots as `1`.

Reviewer: To create this pool is simple.

![](../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-14_at_11.05.02_AM.png)

Will add the name of the pool to your task definition below and that's it.

## Step 7: Build Your Pipeline in Airflow

>   > *If you are familiar with the foundational concepts of Airlfow, you can skip to the next subsection for the `DAG Workflow`*

Reviewer: Add a period at the end of this

> Airflow pipeline design rests on the foundational structures of `Hooks` and `Operators`. Hooks facilitate connections and repository-type actions, whereas Operators use hooks in combinations with domain-specific logic and error handling in order to accomplish a task. Airflow comes pre-packaged with many of these hooks and operators, all built to work with services we use regularly. For example, `BashOperator` allows you to run a bash command directly inline. A simple task using a this operator might look like:

Reviewer: Capitalize Hooks and Operators throughout for consistency.

```python
hello_world = BashOperator(
    task_id='get_schema',
    bash_command='echo "hello, world"'
)
```

Getting more advanced, a `PostgresOperator` executes SQL code in a specific PostgreSQL database using a `PostgresHook` to establish the connection, while templating your command and implementing a `run` function to execute it. A task involving this operator might look like:

```python
insert_row = PostgresOperator(
    task_id='insert_row',
	  postgres_conn_id='default_conn_id',
    sql='INSERT INTO table_1 VALUES (%s, %s, %s)',
    parameters=(1, 'a very important string', datetime.now())
)
```

### DAG Workflow

> Because of the great flexibility of Airflow, you can also use Hooks to create custom Operators, just as well as you can create custom hooks. In this post, we will create a custom `CloudSqlCsvExportOperator` using an existing `CloudSqlHook`, a hook designed to facilitate several types of operations on CloudSQL instances. First, we will use the `BashOperator` to simulate our workflow and check dependencies.

Reviewer: Same way, standardize on capitalizing Hooks and Operators.

Author: going with lowercase, makes more sense

                  "uri": f"gs://{self.gcs_bucket}/{self.gcs_keypath}",
                  "databases": [f"{self.database_name}"], 
                  "offload": self.offload, # True for serverless (takes longer but offloads - use true for historical extract)
                  "csvExportOptions":
                   {
                       "selectQuery": f"{self.export_sql}"
                   }
                }
            }
            hook.export_instance(instance=self.database_instance,
                                 body=body, 
                                 project_id=self.project_id)
    ```

2. Save your new custom operator. Note how some the parameters in the API call indicate this is configurable. You can now use the same code with a different values for `gcs_bucket`, `gcs_keypath`, `export_query`, `project_id`, and `gcp_connection`, making this operator reusable for any variation of this workflow. You can even modify it to fulfill a different component of the same workflow. 

## Step 9: Implement Watermarking and Execution Dates

In data synchronization, a [watermark can be defined as](https://en.wikipedia.org/wiki/Watermark_(data_synchronization)):

> an object of a predefined format which provides a point of reference for two systems/datasets attempting to establish delta/incremental synchronization; any object in the queried data source which was created, modified/changed, and/or deleted after the watermark value was established will be qualified as "above watermark" and could/should be returned to a delta-querying partner

Reviewer: Capitalize this first letter and throw a period at the end here. Also, we should throw quotes around this (and maybe blockquote it?)

Author: good catch. that was a notion formatting drop

Change data in the concept of time-series or events can be looked at by the associated events and their `event_timestamp`. Therefore, we can use event timestamps in increments as high and low watermarks to establish delta synchronizations.

 > *Note: Transactional or non-time series change data can be included here by adding soft deletes (`deleted_at` fields) then watermarking by `deleted_at` and `updated_at` timestamps as per the granularity of timing necessary.*

Airflow has out of the box event timestamps to help with this. When used in combination with the aforementioned source `event_timestamps`, they become extremely helpful in watermarking and partitioning the extracted data. In other words: For all of our task runs, Airflow hands us a value that we can dynamically assign watermarks with and use in our queries. In particular, the `execution_date` is discussed and used here. 

> The `execution_date` is a `pendulum.Pendulum` object available via a jinja templated macro as `{{ execution_date }}`. It's defined as a standard time relative to that which the DAG was scheduled to run at, independent of when it started or was re-ran.

Reviewer: "was re-run"

    │       ├── cloud_sql_cdc
    │       │   └── export_csv
    │       │       ├── table_1.sql
    │       │       ├── table_2.sql
    │       │       ├── table_3.sql
    │       │       ├── table_4.sql
    │       │       ├── get_schema.sql
    ```

 2. Add the following to each `table_{num}.sql` file you created, changing the table name to your own, accordingly:
    ```sql
    {%- set low_watermark =  execution_date.subtract(hours=1) -%}
    SELECT * FROM some_schema.table_1
    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
    AND event_timestamp AT TIME ZONE 'UTC' < '{{ execution_date }}'
    ```

3. Combine all tables into a single schema CSV by adding the following to the  `get_schema.sql` (and changing `some_schema`):
    ```sql
    {% set tables = params.tables %}
    SELECT 
      table_name,
      column_name,
      data_type,
      ordinal_position
    FROM information_schema.columns 
    WHERE table_schema = 'some_schema' 
    AND table_name in (
      {%- for name in tables -%}
        {%- if loop.first -%}
        '{{ name }}'
        {%- else -%}
        ,'{{ name }}'
        {%- endif -%}
      {%- endfor -%}
    )
    ```

>     The `ordinal_position` field is how you will be able to assign column names to header fields in the CSV file. The `params.tables` assignment will come in the DAG file. Note, it is also possible to join tables here (in the need for immediate denormalization) by a `UNION SELECT` on this to the join query with an assigned table name. Now we're able to include functionality in our Airflow tasks.

Reviewer: Think this might end up throwing some markdown errors because only the first line is tabbed out, but check it out in your preview.

              template_searchpath=TEMPLATES_DIR, # Add this**
              start_date=datetime(2020, 12, 29, 0, 0, 0),
              default_args=default_args) as dag:
    ```

    With that, we can simply pass the `.sql` file path to our operators instead of the entire query.  

3. Update the `get_schema` task to show the following:

    ```python
    gcs_schema_keypath = (
        f"{gcs_root}/"
        "schemas/"
        "{{execution_date.year}}/"
        "{{execution_date.month}}/"
        "{{execution_date.day}}/"
        f"{db_name}/"
        f"{db_name}_schema_"
        "{{ts_nodash}}.csv"
    )

    get_schema = CloudSqlCsvExportOperator(
        task_id = 'get_schema',
        gcp_conn_id = gcp_conn_id,
        database_instance = db_instance_id,
        project_id = gcp_project_id,
        database_name = db_name,
        gcs_bucket = gcs_bucket,
        gcs_keypath = gcs_schema_keypath,
        offload = offload_export,
        params = {
            "tables": tables
        },
        export_sql = 'cloud_sql_cdc/export_csv/get_schema.sql',
        pool = 'cloudsql_operations',
        trigger_rule = 'one_success'
    )
    ```

>     > *Notice here how we also used `execution_date` to construct the GCS keypath. You'll see another macro there as well `{{ts_nodash}}`, which is `execution_date.isoformat()` with special characters and spaces removed, useful for time-based object keys.*

Reviewer: "which is the equivalent of"

Author: Good catch. This one is always confusing too. I feel like it's "which is the equivalent to" here because it's used as something that has value

Reviewer 2: When equivalent is used as a noun, you use "of". When it's an adjective, you use "to".

For example:
"This is the equivalent of that"
vs.
"This is equivalent to that"

Author: thank you

When a DAG starts at its scheduled time, it begins at the end of the interval chosen. So for `schedule_interval='@hourly'` and an `execution_date` of `2021-01-14T13:00:00+00:00`, the DAG would actually be starting at `2021-01-14T14:00:00+00:00`. 

For a visual example, the image below shows the task tree of a particular DAG in which this concept can be observed. The first line, `Run: 2021-01-19, 21:00:00 UTC`, shows the `execution_date` from an hourly scheduled DAG, set as `schedule_interval='@hourly'`. Notice `Started: 2021-01-19T22:00:16xxx` is an hour (a single interval) after the schedule date. That is, it starts one `schedule_interval` later than the `execution_date`.

<span style="display:block;text-align:center;">
  <img src="../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-19_at_2.48.37_PM.png" width=400> 
</span>

> The `start_date` you set in the DAG definition follows this logic. If you choose:

Reviewer: "...also follows this logic."

```python
start_date = DateTime(2021, 1, 14, 13, 0, 0),
schedule_interval = '@hourly',
catchup = False
```

> you would see the first run at `2021-01-14T14:00:00+00:00` if the current date were before that and before the `start_date`. If the current calendar date was after the `start_date`, you would see the first run start at the beginning of the most recent hour. 

Reviewer: Capitalize the You.

Reviewer: Actually, would change this sentence to be parallel to the following one.

"If the current calendar date were before...., you would see the first run at xxx. Conversely, if the current calendar date were after..."

Author: Agreed it's tough. There's 3 dates happening though: The current calendar date, the start_date, and the execution_date.

    ```

    which you would see renders to

    ```bash
    # ----------------------------------------------------------
    # property: export_sql
    # ----------------------------------------------------------
    SELECT * FROM some_schema.table_1
    WHERE event_timestamp AT TIME ZONE 'UTC' >= '2021-01-14T13:00:00+00:00' AT TIME ZONE 'UTC'
    AND event_timestamp AT TIME ZONE 'UTC' < '2021-01-14T14:00:00+00:00'
    ```

    getting you changes in the previous hour. 

2. Make the following changes to the rest of your `.sql` files.

3. In your DAG definition, adjust to your interval of choice. Suppose you want to make this run every 30 minutes. The `schedule_interval` parameter also accepts a crontab value. So in the DAG definition, that would be `schedule_interval = '* /30 * * * *'`, and your templated SQL would appear:

    ```bash
    {%- set low_watermark =  execution_date -%}
    {%- set high_watermark =  execution_date.add(minutes=30) -%}
    SELECT * FROM some_schema.table_1
    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
    AND event_timestamp AT TIME ZONE 'UTC' < '{{ high_watermark }}'
    ```

>     In a future post, I will show you how to set these dynamically, through the UI, as to allow interval changes and full scans when needed.

Reviewer: Don't think we need the commas here around `through the UI`.

"In a future post, I will show you how to set these dynamically through the UI to allow interval changes and full scans."

                f"{table}_"
                "{{ts_nodash}}.csv"
            )

            export_table = CloudSqlCsvExportOperator(
                task_id = f'export_{table}',
                gcp_conn_id = gcp_conn_id,
                database_instance = db_instance_id,
                project_id = gcp_project_id,
                database_name = db_name,
                gcs_bucket = gcs_bucket,
                gcs_key = gcs_raw_keypath,
                offload = offload_export,
                export_sql = f'cloud_sql_cdc/export_csv/{table}.sql',
                pool = 'cloudsql_operations',
                trigger_rule='one_success'
            )
            get_schema >> export_table >> complete_export  
            
        kickoff_dag >> start_export >> get_schema
        complete_export >> complete_dag
    ```
    Now, we have a DAG that extracts change data from our tables at an hourly interval as soon as that interval completes. We can now deploy this to production.

## Step 14: Deploy to Production

Once you commit your changes to git and merge to branches as needed, you are ready to deploy this to production.

### **Create an Airflow Deployment**

> To create an Airflow Deployment on Astronomer, log into [Astronomer Cloud](https://app.gcp0001.us-east4.astronomer.io/), open your Workspace, and click **New Deployment**.

Reviewer: Change this link and language up a bit- we currently don't allow self-service for new signups.

"To create an Airflow Deployment on Astronomer, start a free trial of [Astronomer Cloud](https://astronomer.io/get-astronomer)..."

Reviewer: ^that link is to https://astronomer.io/get-astronomer

Author: These are not my words. I took it from the other page https://www.astronomer.io/docs/cloud/stable/deploy/deploy-cli#step-1-create-an-airflow-deployment

This is a good use case for markdown modules. Then we can change one and have it update everywhere.

Author: Feel free to change this one or resolve as needed. I don't know

Reviewer 2: I think we just need an additional step telling people to create an Astronomer account. If that's the case, we can appropriate the language from this step of the [Cloud quickstart](https://www.astronomer.io/docs/cloud/stable/get-started/quickstart#step-1-start-your-trial):

![https://assets2.astronomer.io/main/docs/deploying-code/v0.23-deployments.png](https://assets2.astronomer.io/main/docs/deploying-code/v0.23-deployments.png)

> ## **Configure Your Airflow Deployment**

Reviewer: Same here

Use the **New Deployment** menu to configure the following:

- **Name**
- **Description** (Optional)
- **Airflow Version**: We recommend using the latest version.
- **Executor**: We recommend starting with Local.

![https://assets2.astronomer.io/main/docs/deploying-code/v0.23-new_deployment-config.png](https://assets2.astronomer.io/main/docs/deploying-code/v0.23-new_deployment-config.png)

When you've finished, click **Create Deployment**.

Once you've initialized your Deployment, give it a few moments to spin up. Afterwards, you'll have access to your Deployment dashboard:

![https://assets2.astronomer.io/main/docs/deploying-code/v0.23-new_deployment-dashboard.png](https://assets2.astronomer.io/main/docs/deploying-code/v0.23-new_deployment-dashboard.png)

From this dashboard, you can:

- Modify your Resources
- Initialize the Airflow upgrade process
- Configure **Email Alerts**
- Access the Airflow UI
- Access your Celery Dashboard (if using CeleryExecutor)
- Delete your Deployment

For more information on deployment configuration, read [Configure an Airflow Deployment on Astronomer](https://www.astronomer.io/docs/cloud/stable/deploy/configure-deployment/).

> ## **Deploy Code from the CLI**

Reviewer: Same here

> ### **a. Authenticate via the Astronomer CLI**

Reviewer: Same Here

Reviewer: Also, does it make sense to number these steps instead of letter them? That may be more intuitive.

To authenticate via the Astronomer CLI, run:

```bash
astro auth login gcp0001.us-east4.astronomer.io
```

> ### **b. Confirm Your Workspace and Deployment**

Reviewer: Same here.

From the Astronomer CLI, you're free to push code to any Airflow Deployment you have access to as long as you have the appropriate deployment-level permissions to do so.

Before you deploy to Astronomer, make sure that the Airflow Deployment you'd like to push code to is within the Workspace you're operating in.

To see the list of Workspaces you have access to, run:

```bash
astro workspace list
```

To switch between Workspaces, run:

```bash
astro workspace switch
```

To see the list of Deployments within a particular Workspace, run:

```bash
astro deployment list
```

For more specific CLI guidelines and commands, read [CLI Quickstart](https://www.astronomer.io/docs/cloud/stable/develop/cli-quickstart/).

> ### **c. Deploy to Astronomer**

Reviewer: Same here.

Finally, make sure you're in the correct Airflow project directory.

When you're ready to deploy your DAGs, run:

```bash
astro deploy
```

This command returns a list of Airflow Deployments available in your Workspace and prompts you to pick one.

> ### d**. Start the DAG**

Reviewer: Same here.

## Overview

The scheduled approach outlined in this post - enabling change data to be updated in near-real time - has the added benefit of built-in data partitioning formats directly from the process. In a distributed data network, using Airflow as the scheduled approach pushes it into the realm of Consistency and High Availability (CA) in the CAP theorem. That is, you can base SLAs on this approach.

Those capturing data in the format of time-series or events will benefit the most with this approach, but it can be used for any data by adding soft deletes to your transactions and defining the time scope to which changes must be recorded. Eventual consistency is achieved in a robust manner by creating an idempotent approach to how the data is extracted and moved. 

> In this series, we'll outline all steps in the process from end to end, from setup to extract. In totality, this will include topics of:

Reviewer 2: ```suggestion
In this series, we'll outline all steps in the process from end to end, from setup to extract. Specifically, this entails:
```

- Setting up security and IAM in GCP
- Calling the CloudSQL Export API
- Setting up Airflow and connecting it to GCP
- Building a DAG in Airflow
- Configuring Watermarking for data extract intervals
- Writing an operator in Airflow to extract the data
- Integrating the operator into the DAG
- Scheduling and coordinating the extract
- Deployment to Production on Astronomer Cloud

In a later post, I will show you how to transform or clean the data then add it to a warehouse as part of the same workflow. We will be using the [Astronomer Platform](https://www.astronomer.io/), as it is the easiest way to get your Apache Airflow instance up and running.

## How it Works

> The base method for extract uses the [CloudSQL Export API](https://cloud.google.com/sql/docs/postgres/import-export/exporting#rest-v1beta4). This runs an operation with a SQL query, sending the results to a Google Cloud Storage (GCS) bucket as declared by you. You could just as easily extract directly from your database if your system and workload allows for such. 

Reviewer 2: ```suggestion
The base method for extract uses the [CloudSQL Export API](https://cloud.google.com/sql/docs/postgres/import-export/exporting#rest-v1beta4). This runs an operation with a SQL query, sending the results to a specified Google Cloud Storage (GCS) bucket. You could just as easily extract directly from your database if your system and workload allows for such. 
```

Author: @jwitz how about , "sending the results to a Google Cloud Storage (GCS) bucket specified by you". I'm trying let them know they have to create the bucket, as they will see in a subsequent step. So they may need permission by an admin to be able to do that.

Reviewer 2: For sure, how about this (eliminates passive language):
```suggestion
The base method for extract uses the [CloudSQL Export API](https://cloud.google.com/sql/docs/postgres/import-export/exporting#rest-v1beta4). This runs an operation with a SQL query, sending the results to a Google Cloud Storage (GCS) that you specify. You could just as easily extract directly from your database if your system and workload allows for such. 
```

> Using the extract API provides a robust and easily configurable security layer controlled by a service account, such that no system outside of CloudSQL is touching your production database. It also provides a "serverless offload" option in the case that you need to guarantee zero load on your database from the extract. 

Reviewer 2: ```suggestion
Using the extract API provides a robust and easily configurable security layer controlled by a service account. This ensures that no system outside of CloudSQL is touching your production database. It also provides a "serverless offload" option in case you need to guarantee zero load on your database from the extract. 
```

Author: @jwitz "In case" feels different. Is there another way we can say it? Saying "in case you need" expresses that option as more of a backup plan instead of a specific use-case. Some people may have such a big extract to do, they won't want to push on their DB, and instead opt to use the serverless offload.

Reviewer 2: Let's keep it simple then:
```suggestion
Using the extract API provides a robust and easily configurable security layer controlled by a service account, such that no system outside of CloudSQL is touching your production database. It also provides a "serverless offload" option if you need to guarantee zero load on your database from the extract. 
```

Author: How about "It also provides a "serverless offload" option in case you need to guarantee zero load on your database from the extract"

## Step 4: Initialize Your Airflow Project

If you are already familiar with creating a new Airflow project, you can skip to the next section. 

1. To create a new project with the Astronomer CLI, in your terminal, simply create a new directory for your project and then use the Astronomer CLI to scaffold out your project.

    ```bash
    mkdir <directory-name> && cd <directory-name>
    ```

    ```bash
    astro dev init
    ```

    This will generate the following files in that directory:

    ```bash
    .
    ├── dags # Where your DAGs go
    │   ├── example-dag.py # An example dag that comes with the initialized project
    ├── Dockerfile # For Astronomer's Docker image and runtime overrides
    ├── include # For any other files you'd like to include
    ├── plugins # For any custom or community Airflow plugins
    ├──airflow_settings.yaml #For your Airflow Connections, Variables and Pools (local only)
    ├──packages.txt # For OS-level packages
    └── requirements.txt # For any Python packages
    ```

> 2. For version consistency at the time of writing, let's manually set the image our `Dockerfile` uses. Open the `Dockerfile`, change the contents to read:

Reviewer 2: ```suggestion
2. For version consistency at the time of writing, let's manually set the image our `Dockerfile` uses. Open the `Dockerfile` and change the contents to read:
```

    ```docker
    FROM quay.io/astronomer/ap-airflow:1.10.14-1-buster-onbuild
    ```

> 3. Next, In the `dags` directory create a new python file with appropriate naming of your choice, such as `change_data_capture.py`. Now (with docker running) start up airflow from the CLI by running:

Reviewer 2: ```suggestion
3. Next, In the `dags` directory create a new python file with appropriate naming of your choice, such as `change_data_capture.py`. Now (with docker running) start up Airflow from the CLI by running:
```

    ```bash
    astro dev start
    ```

4. Once the containers are built and the webserver starts, you can access the UI at  `[localhost:8080](http://localhost:8080)` and login with `admin:admin`.

## Step 5: Create a Connection to GCP

Now we need to add the service account JSON from GCP to Airflow to create a connection. To do so: 

1. In the Airflow UI, open the `Admin -> Connections` tab and create a new connection with the `Connection Type` of `Google Cloud Platform`. 
2. In the `Scopes` field, add the the following scope:

    ```bash
    https://www.googleapis.com/auth/cloud-platform
    ```

3. In the `Keyfile JSON` field, paste the Keyfile JSON you copied from Step 2. 

    ![Connections](../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-13_at_4.25.24_PM.png)

Keep the `Conn Id` handy. You will use that in your Airflow DAG to reference this connection. When you're ready, save this connection to finalize your access management. 

## Step 6: Configure Pooling

> The next configuration you'll set is a pool. From the [Airflow docs](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=pool#pools), `Airflow pools can be used to limit the execution parallelism on arbitrary sets of tasks`, and therefore, using them will enable you to create other parallel jobs while still guaranteeing that your export API tasks will run in serial. 

Reviewer 2: ```suggestion
The next configuration you'll set is a pool. From the [Airflow docs](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=pool#pools), "Airflow pools can be used to limit the execution parallelism on arbitrary sets of tasks". Therefore, you can use them to create other parallel jobs while still guaranteeing that your export API tasks will run in serial. 
```

> To create this pool is simple. Just navigate in the top bar to `Admin -> Pools`, then hit the `+` sign to add a new record. Choose a pool name and description to your liking, but choose the number of slots as `1`.

Reviewer 2: ```suggestion
To create this pool, simply go to the top bar to `Admin -> Pools`, then hit the `+` sign to add a new record. Choose a pool name and description to your liking, but choose the number of slots as `1`.
```

Author: @jwitz How about "To create this pool, simply navigate in the top bar to"

Reviewer 2: I'm generally hostile to the word "navigate" since it has a lot of connotations and is much longer than "go". I would simply say "To create this pool, go to `Admin -> Tools`  in the Airflow UI and click the `+` sign". This language is a little more accessible and future proof as well - who know how long those buttons will be in a top bar!

![Pools](../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-14_at_11.05.02_AM.png)

> Will add the name of the pool to your task definition below and that's it.

Reviewer 2: ```suggestion
Now all we have to do is define the name of the pool in our Airflow DAG, which we'll build in the next step.
```

## Step 7: Build Your Pipeline in Airflow

  > *If you are familiar with the foundational concepts of Airflow, you can skip to the next subsection for the `DAG Workflow`.*

> Airflow pipeline design rests on the foundational structures of `hooks` and `operators`. Hooks facilitate connections and repository-type actions, whereas operators use hooks in combinations with domain-specific logic and error handling in order to accomplish a task. Airflow comes pre-packaged with many of these hooks and operators, all built to work with services we use regularly. For example, `BashOperator` allows you to run a bash command directly in-line. A simple task using a this operator might look like:

Reviewer 2: ```suggestion
Airflow pipeline design rests on the foundational structures of `hooks` and `operators`. Hooks facilitate connections and repository-type actions, whereas operators use hooks in combination with domain-specific logic and error handling in order to accomplish a task. Airflow comes pre-packaged with many of these hooks and operators, all built to work with services we use regularly. For example, `BashOperator` allows you to run a bash command directly in-line. A simple task using a this operator might look like:
```

```python
hello_world = BashOperator(
    task_id='get_schema',
    bash_command='echo "hello, world"'
)
```

Getting more advanced, a `PostgresOperator` executes SQL code in a specific PostgreSQL database using a `PostgresHook` to establish the connection, while templating your command and implementing a `run` function to execute it. A task involving this operator might look like:

```python
insert_row = PostgresOperator(
    task_id='insert_row',
	  postgres_conn_id='default_conn_id',
    sql='INSERT INTO table_1 VALUES (%s, %s, %s)',
    parameters=(1, 'a very important string', datetime.now())
)
```

### DAG Workflow

Because of the great flexibility of Airflow, you can also use hooks to create custom operators, just as well as you can create custom hooks. In this post, we will create a custom `CloudSqlCsvExportOperator` using an existing `CloudSqlHook`, a hook designed to facilitate several types of operations on CloudSQL instances. First, we will use the `BashOperator` to simulate our workflow and check dependencies.

> In the DAG workflow, it is extremely advantageous to also create generic tasks that serve only as checkpoints in the workflow. This facilitates best practices like idempotency, task grouping, workflow readability, and traceability. This is accomplished by a stand-in operator called the `DummyOperator`. In the same effort, the `DummyOperator` is an excellent way to first build your workflow and visualize task dependencies. Combining this with a `BashOperator` allows us to add a "Hello, world" level of functionality to this. This is what we will do first.

Reviewer 2: ```suggestion
In the DAG workflow, it is extremely advantageous to also create generic tasks that serve only as checkpoints in the workflow. This facilitates best practices like idempotency, task grouping, workflow readability, and traceability. This is accomplished by a stand-in operator called the `DummyOperator`. Additionally, the `DummyOperator` is an excellent way to first build your workflow and visualize task dependencies. Combining this with a `BashOperator` allows us to add a "Hello, world" level of functionality to this. To implement the `DummyOperator`:
```

Author: @jwitz I totally re-wrote this. Let me know what you think.

In the DAG workflow, it is extremely advantageous to also create generic tasks that serve only as checkpoints in the workflow. This facilitates best practices like idempotency, task grouping, workflow readability, and traceability. This is accomplished by a stand-in operator called the `DummyOperator`. Building your whole DAG first with all `DummyOperator` tasks is an excellent way to visualize task dependencies and control the flow of tasks. As a next step, combining this with a `BashOperator` allows us to add a "Hello, world" level of functionality to this and test the flow of the DAG itself. To do this:

Reviewer 2: Using some more metaphorical language to communicate your point

"Outlining your DAG using DummyOperator tasks is an excellent way to visualize task dependencies and control the flow of tasks. As a next step, we can use a BashOperator to add a "Hello, world" level of functionality and test the flow of the DAG. To do this:"

From what I understand, we're essentially building an outline. I think it's important to get that word in there as a cornerstone for what you're trying to communicate.

        get_schema = BashOperator(
            task_id='get_schema',
            bash_command='echo "getting schema"',
            trigger_rule='one_success',
            pool = 'cloudsql_operations'
        )
         
        complete_dag = DummyOperator(task_id='complete_dag')

        for table in tables:

            export_table = BashOperator(
                task_id=f'export_{table}',
                bash_command=f'echo "exporting {table}"',
                trigger_rule='one_success',
                pool = 'cloudsql_operations'
            )

            get_schema >> export_table >> complete_export  
            
        kickoff_dag >> start_export >> get_schema
        complete_export >> complete_dag
    ```

    This workflow is simply creating a set of dependent tasks and instructing them to run in a pool with a single slot, thereby forcing them to run one at a time. Save the file. The bash commands will appear in the tasks logs when run. 

2. Now, open the Airflow UI, click on your DAG in the list, then navigate to the `Graph View` . It should look something this:

    ![Graph View](../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-20_at_10.57.19_AM.png)

> 3. You can now switch `Off` to `On`  and choose `Trigger DAG` to see this workflow run then verify each export task runs one at a time by looking at the `started` time for each. 

Reviewer 2: ```suggestion
3. Switch `Off` to `On`  and choose `Trigger DAG` to see this workflow run, then verify each export task runs one at a time by looking at the `started` time for each. 
```

<span style="display:block;text-align:center;">![](../assets/cdc-cloudsql-1/cdc_cloudsql_airflow_2021-01-20_at_10.58.55_AM.png)</span>

> If it doesn't run, check your start date is set to the past, not the future. If your DAG is not in the list and doesn't appear after a minute, you may have a syntax error in your python file. If it's good, we turn the DAG back to `Off` and start building in functionality.

Reviewer 2: ```suggestion
If it doesn't run, check that your start date is set to the past, not the future. If your DAG is not in the list and doesn't appear after a minute, you may have a syntax error in your python file. 

Once the DAG is running as expected, we can turn the DAG back to `Off` and start building in functionality.
```

## Step 8: Build the CloudSqlCsvExportOperator

In the `plugins` directory, create a new file with the path:

```bash
├── plugins
│   ├── gcp_custom
│   │   ├── operators
│   │   │   └── gcp_cloudsql_export_operator.py
```

> As long as you add this in the plugins directory, the path can be arbitrary, but this is a recommended format for modularization. 

Reviewer 2: ```suggestion
The path is arbitrary once you add this in your `plugins` directory, but this is a recommended format for modularization. 
```

            body = {
                "exportContext":
                {
                  "fileType": f"{self.file_type}",
                  "uri": f"gs://{self.gcs_bucket}/{self.gcs_keypath}",
                  "databases": [f"{self.database_name}"], 
                  "offload": self.offload, # True for serverless (takes longer but offloads - use true for historical extract)
                  "csvExportOptions":
                   {
                       "selectQuery": f"{self.export_sql}"
                   }
                }
            }
            hook.export_instance(instance=self.database_instance,
                                 body=body, 
                                 project_id=self.project_id)
    ```

2. Save your new custom operator. Note how some the parameters in the API call indicate this is configurable. You can now use the same code with a different values for `gcs_bucket`, `gcs_keypath`, `export_query`, `project_id`, and `gcp_connection`, making this operator reusable for any variation of this workflow. You can even modify it to fulfill a different component of the same workflow. 

## Step 9: Implement Watermarking and Execution Dates

> In data synchronization, a [watermark can be defined as](https://en.wikipedia.org/wiki/Watermark_(data_synchronization)):

Reviewer 2: ```suggestion
In data synchronization, a [watermark](https://en.wikipedia.org/wiki/Watermark_(data_synchronization))  can be defined as:
```

    │       │   └── export_csv
    │       │       ├── table_1.sql
    │       │       ├── table_2.sql
    │       │       ├── table_3.sql
    │       │       ├── table_4.sql
    │       │       ├── get_schema.sql
    ```

2. Add the following to each `table_{num}.sql` file you created, changing the table name to your own, accordingly:
    
    ```sql
    {%- set low_watermark =  execution_date.subtract(hours=1) -%}
    SELECT * FROM some_schema.table_1
    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
    AND event_timestamp AT TIME ZONE 'UTC' < '{{ execution_date }}'
    ```

3. Combine all tables into a single schema CSV by adding the following to the  `get_schema.sql` (and changing `some_schema`):
    
    ```sql
    {% set tables = params.tables %}
    SELECT 
      table_name,
      column_name,
      data_type,
      ordinal_position
    FROM information_schema.columns 
    WHERE table_schema = 'some_schema' 
    AND table_name in (
      {%- for name in tables -%}
        {%- if loop.first -%}
        '{{ name }}'
        {%- else -%}
        ,'{{ name }}'
        {%- endif -%}
      {%- endfor -%}
    )
    ```

>     The `ordinal_position` field is how you will be able to assign column names to header fields in the CSV file. The `params.tables` assignment will come in the DAG file. Note, it is also possible to join tables here (in the need for immediate denormalization) by a `UNION SELECT` on this to the join query with an assigned table name. Now we're able to include functionality in our Airflow tasks.

Reviewer 2: ```suggestion
    The `ordinal_position` field is how you will be able to assign column names to header fields in the CSV file. The `params.tables` assignment will come in the DAG file. With that in place, we're now able to include functionality in our Airflow tasks.

> **Note:** It's also possible to join tables here (in the need for immediate denormalization) by running a `UNION SELECT` on this to the join query with an assigned table name.
```

4. Change the  `export_table` tasks to show the following:

    ```python
    for table in tables:

        gcs_raw_keypath = (
            f"{gcs_root}/"
            "raw/"
            f"{table}/"
            "{{execution_date.year}}/"
            "{{execution_date.month}}/"
            "{{execution_date.day}}/"
            f"{table}_"
            "{{ts_nodash}}.csv"
        )

        export_table = CloudSqlCsvExportOperator(
            task_id = f'export_{table}',
            gcp_conn_id = gcp_conn_id,
            database_instance = db_instance_id,
            project_id = gcp_project_id,
            database_name = db_name,
            gcs_bucket = gcs_bucket,
            gcs_keypath = gcs_raw_keypath,
            offload = offload_export,
            export_sql = f'cloud_sql_cdc/export_csv/{table}.sql',
            pool = 'cloudsql_operations',
            trigger_rule = 'one_success'
        )

        get_schema >> export_table >> complete_export
    ```

>     As mentioned above, all source code here is [available for download on github](https://github.com/astronomer/cdc-cloudsql-airflow-demo). Now it's time to check that everything is set up by using the render tool.

Reviewer 2: ```suggestion
    As mentioned above, all source code here is [available for download on github](https://github.com/astronomer/cdc-cloudsql-airflow-demo). Now it's time to check that everything works by using the render tool.
```

## Step 11: Render Template Check

> In your terminal, you can use the Astro CLI to render all templating using the following format:

Reviewer 2: ```suggestion
In your terminal, you can use the Astro CLI to render all templating using the following command:
```

Author: @jwitz how about "templating using a command with the following format:"?
We don't want them to copy-paste that line.

Reviewer 2: It's standard across our docs to use pointed brackets as placeholders in a command, I think that's self-explanatory here.

```bash
astro dev run render <dag-id> <task-id> 'execution_date'
```

Based on this example, we can run:

```bash
astro dev run render change_data_capture_cloud_sql export_table_1 '2021-01-14 13:00:00'
```

Afterwards, we should see our templating print out:

```bash
# ----------------------------------------------------------
# property: export_sql
# ----------------------------------------------------------
SELECT * FROM some_schema.table_1
WHERE event_timestamp AT TIME ZONE 'UTC' >= '2021-01-14T12:00:00+00:00' AT TIME ZONE 'UTC'
AND event_timestamp AT TIME ZONE 'UTC' < '2021-01-14T13:00:00+00:00'

# ----------------------------------------------------------
# property: gcs_bucket
# ----------------------------------------------------------
your_gcs_bucket

# ----------------------------------------------------------
# property: gcs_keypath
# ----------------------------------------------------------
your_gcs_bucket/cloudsql_export/raw/table_1/2021/1/14/table_1_20210114T130000.csv
```

> We see our SQL Query, our bucket, and the full `gcs_keypath` fully rendered based on our `execution_date` chosen. Try the same for `get_schema`. When complete, the success of such signal this DAG is now ready to be run. 

Reviewer 2: ```suggestion
We see our SQL Query, our bucket, and the full `gcs_keypath` fully rendered based on our `execution_date` chosen. Try the same for `get_schema`. If this templating renders correctly, then the DAG is now ready to be run. 
```

    SELECT * FROM some_schema.table_1
    WHERE event_timestamp AT TIME ZONE 'UTC' >= '2021-01-14T13:00:00+00:00' AT TIME ZONE 'UTC'
    AND event_timestamp AT TIME ZONE 'UTC' < '2021-01-14T14:00:00+00:00'
    ```

    getting you changes in the previous hour. 

2. Make the following changes to the rest of your `.sql` files.

3. In your DAG definition, adjust to your interval of choice. Suppose you want to make this run every 30 minutes. The `schedule_interval` parameter also accepts a crontab value. So in the DAG definition, that would be `schedule_interval = '* /30 * * * *'`, and your templated SQL would appear:

    ```bash
    {%- set low_watermark =  execution_date -%}
    {%- set high_watermark =  execution_date.add(minutes=30) -%}
    SELECT * FROM some_schema.table_1
    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
    AND event_timestamp AT TIME ZONE 'UTC' < '{{ high_watermark }}'
    ```

    In a future post, I will show you how to set these dynamically through the UI as to allow interval changes and full scans when needed.

### Step 13: Intro to Dynamic Configs with Airflow Variables

> Airflow Variables are simple key-value fields that can be added through the Airflow UI and parsed in your DAG. They are useful for adding dynamic configuration values to your DAG file. With this, it's extremely important to realize every loop of the scheduler will parse it, making a connection to the database each time. To make this efficient, it's recommended to add a single JSON-serialized variable, then access individual values as items in a dictionary. To convert our hard-coded strings:

Reviewer 2: ```suggestion
Airflow Variables are simple key-value fields that can be added through the Airflow UI and parsed in your DAG. They are useful for adding dynamic configuration values to your DAG file. With this, it's extremely important to realize every loop of the scheduler will parse it, meaning that every loop of the scheduler makes a connection to the database. To make this efficient, it's recommended to add a single JSON-serialized variable, then access individual values as items in a dictionary. To convert our hard-coded strings:
```

Author: @jwitz I changed this to "With this, it's extremely important to realize that on every loop of the scheduler a query is made to the database for each variable."

Reviewer 2: "With this, it's important to note that for every loop of the scheduler, a query is made to the database for each variable"
